src/qutip_tensorflow/core/data/expect.py

```diff
@@ -0,0 +1,75 @@
+import qutip
+from math import sqrt
+from .tftensor import TfTensor
+import warnings
+
+with warnings.catch_warnings():
+    warnings.filterwarnings("ignore", category=DeprecationWarning)
+    import tensorflow as tf
+
+
+__all__ = ["expect_tftensor", "expect_super_tftensor"]
+
+
+def _check_shape_expect(op, state):
+    if (
+        op.shape[0] != op.shape[1]  # Op square matrix.
+        or not (
+            state.shape[0] == state.shape[1]  # state is square matrix
+            or state.shape[1] == 1  # state is ket
+        )
+        or op.shape[1] != state.shape[0]  # Not possible op@state
+    ):
+        raise ValueError(
+            f"incompatible matrix shapes for op {op.shape} and state {state.shape}"
+        )
+
+
+def _check_shape_super(op, state):
+    if state.shape[1] != 1:
+        raise ValueError(
+            "expected a column-stacked matrix but state input matrix"
+            + f"has shape {state.shape}"
+        )
+    if op.shape[1] != state.shape[0]:
+        raise ValueError(f"incompatible shapes op:{op.shape} and state:{state.shape}")
+    if op.shape[0] != op.shape[1]:
+        raise ValueError(
+            "Expect_super only works for square op matrices. The"
+            f"provided op has shape: {op.shape}"
+        )
+
+
+def expect_ket_tftensor(op, state):
+    return tf.reduce_sum(tf.math.conj(state._tf) * (op._tf @ state._tf))
```

Comment:
I'm surprised there's no built-in `tf.dot` or `tf.vdot`.  Do you know how many full passes through the tensors this does?  Or are we still just building up the operation graph at this point?

Reply:
The code is executed eagerly  so it first performs the matmul operation, then the conj operation and then the elementwise_mul operations. Are you worried that the operation may create a lot of unnecessary tensors? If thats the case, so am I, but there does not seem to be native "reduce" operation that fits our needs here. The work around could be einsum but I am not 100% sure how this will work or if it will even provide any speed-up. Another option is to use a tf.function to explicitly build a graph for the operation. Lets discuss it on today's meeting.

Reply:
`einsum('ab,bc,cd->ad', state._tf.conjugate(), op._tf, state._tf)` should be the same thing, with only one temporary structure, but `tensorflow.function` is probably the correct way to go (subject to benchmarks).

Reply:
So i did a few benchamarks and the results are quite disapointing.
The code
```python
def expect(op, state):
    return tf.reduce_sum(tf.math.conj(state) * (op @ state))

# Random tf.Tensor
herm = qutip.rand_herm(2000).to('tftensor')
herm = herm.data._tf 


# Setting jit_compile to true is even slower 
expect_jit = tf.function(expect, jit_compile=False)  
# Compile the function once to avoid compilation time in bechmark
expect_jit(herm, herm)

%%timeit  # 80ms
tf.einsum('ab,bc,ca', tf.math.conj(herm), herm, herm).numpy()

%%timeit  # 91
expect_jit(herm, herm).numpy()

%%timeit  # 81 ms
expect(herm, herm).numpy()
``` 
The code was run in colab with a GPU.

I think it is best to just leave the function as it is for the moment. I may try in the future to squeeze some performance using `tf.function`. I do not know why but im never able to improve the performance of my code using jit. I also tried it with numba in the past and had the same luck as just now.

Reply:
To get big improvements with Numba, you generally have to change coding style a little - if you're already writing well vectorised Numpy code , Numba won't give you so much. You get more improvements if there's big Python loops and conditionals, which isn't the natural way of writing Numpy code (because it's slow as anything without Numba).

Probably here the function is sufficiently simple that the overhead of `tf.function` isn't enough to beat the allocation and copying cost. It may even be that the graphed operation doesn't even remove the copy in this case. Good job looking into it.